extension/textext/__init__.py

```diff
@@ -416,27 +389,35 @@ def do_convert(self, text, preamble_file, user_scale_factor, old_svg_ele, alignm
                     # no version attribute is provided by Inkscape :-(
                     pass
 
-                # -- Copy style
+                # Place new node in document
                 if old_svg_ele is None:
                     with logger.debug("Adding new node to document"):
-                        root = self.document.getroot()
-                        width = self.svg.unittouu(self.get_document_width())
-                        height = self.svg.unittouu(self.get_document_height())
+                        # Place new nodes in the center of the document and scale them according to user request
+                        import inkex.units
+
+                        # Get center of the document (in user units, e.g. mm in Europe)
+                        doc_cx, doc_cy = self.svg.get_center_position()
+
+                        # Get center of node. Its units are in pt, so convert them to the units used in the document
+                        node_cx, node_cy = tt_node.get_center_position()
+                        node_cx = node_cx*self.svg.unittouu("1pt")
+                        node_cy = node_cy*self.svg.unittouu("1pt")
 
-                        x, y, w, h = tt_node.bounding_box()
                         tt_node.transform = tt_node.transform * \
-                                            Transform(translate=(-x + width / 2 - w / 2, -y + height / 2 - h / 2)) * \
+                                            Transform(translate=(doc_cx - node_cx, doc_cy - node_cy)) * \
                                             Transform(scale=scale_factor)
+
```

Comment:
I tested it with number of document units and scales and it doesn't seem to do right thing. 

1) `node.get_center_position()` returns in user units already
2) `svg.get_center_position()` returns *view* center, not document (bad naming in `inkex`, I'll file an issue)
3) scale was applied *before* compensation of the node width

The following snippet does the trick. Note: now node is placed in middle of view, not in center of document as before. I think it's even better

```python
                        # Place new nodes in the view center and scale them according to user request

                        from inkex.transforms import Vector2d

                        node_center = Vector2d(tt_node.get_center_position())
                        view_center = self.svg.get_center_position()

                        tt_node.transform = (Transform(translate=view_center) *   # place at view center
                                             Transform(scale=scale_factor) *      # scale 
                                             Transform(translate=-node_center) *  # place node at origin
                                             tt_node.transform                    # use original node transform
                                             )
```

Reply:
Thanks for the careful review. My mistake was to test only with a fully visible A4 page centered in the screen after pressing "5" so center of view is equal to center of page. And scaling before moving, you are right, of course.

However, I cannot reproduce your statement

>     1. `node.get_center_position()` returns in user units already

Under both, Linux and Windows the node data is returned in pt in by environment.

Furthermore, despite what I am setting under Inkscape - Document properties, user units are always in mm. Under Edit - Preferences I do not find anything. What I am missing, where is my misunderstanding?

Reply:
> Note: now node is placed in middle of view, not in center of document as before. I think it's even better

This is definitely a great enhancement!

Reply:
> 1. `node.get_center_position()` returns in user units already

I cite here doc string of the function. I tried different units in document properties. The size of output was same and matched text objects created by inkscape. 

Do you observe any misbehavior? As long as we use one and the same units it doesn't matter which one.

Reply:
> This is definitely a great enhancement!

It should be reflected in docs and change list. Not necessarily as part of this PR, but it will be better.

Reply:
Doing an ``etree.dump(root)`` after import of the generated node svg at

https://github.com/textext/textext/blob/ccf023ed75b77a210c6d005f8998df881001c306/extension/textext/__init__.py#L637-L642

one obtains (compiled LaTeX snippet: ``$\times$``)

```xml
<svg xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:cc="http://creativecommons.org/ns#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd" xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" sodipodi:docname="tmp.pdf" id="svg17" version="1.2" viewBox="0 0 4.8125 4.796875" height="4.796875pt" width="4.8125pt">
  <metadata id="metadata21">
    <rdf:RDF>
      <cc:Work rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type rdf:resource="http://purl.org/dc/dcmitype/StillImage"/>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <sodipodi:namedview id="namedview19" inkscape:window-height="480" inkscape:window-width="640" inkscape:pageshadow="2" inkscape:pageopacity="0" guidetolerance="10" gridtolerance="10" objecttolerance="10" borderopacity="1" inkscape:document-rotation="0" bordercolor="#666666" pagecolor="#ffffff"/>
  <defs id="defs10">
    <g id="g8">
      <symbol id="glyph0-0" overflow="visible">
        <path inkscape:connector-curvature="0" id="path2" d="" style="stroke:none"/>
      </symbol>
      <symbol id="glyph0-1" overflow="visible">
        <path inkscape:connector-curvature="0" id="path5" d="M 3.875,-2.765625 1.890625,-4.75 C 1.765625,-4.875 1.75,-4.890625 1.671875,-4.890625 1.5625,-4.890625 1.46875,-4.8125 1.46875,-4.6875 c 0,0.0625 0.015625,0.078125 0.125,0.1875 l 2,2 -2,2.015625 c -0.109375,0.109375 -0.125,0.125 -0.125,0.1875 0,0.125 0.09375,0.203125 0.203125,0.203125 0.078125,0 0.09375,-0.015625 0.21875,-0.140625 L 3.875,-2.21875 l 2.0625,2.0625 c 0.015625,0.015625 0.078125,0.0625 0.140625,0.0625 0.125,0 0.203125,-0.078125 0.203125,-0.203125 0,-0.015625 0,-0.046875 -0.03125,-0.109375 C 6.234375,-0.421875 4.65625,-1.984375 4.15625,-2.5 l 1.828125,-1.8125 c 0.046875,-0.0625 0.203125,-0.1875 0.25,-0.25 0,-0.015625 0.046875,-0.0625 0.046875,-0.125 0,-0.125 -0.078125,-0.203125 -0.203125,-0.203125 C 6,-4.890625 5.96875,-4.859375 5.859375,-4.75 Z m 0,0" style="stroke:none"/>
      </symbol>
    </g>
  </defs>
  <g transform="translate(-150.18076,-129.87437)" id="surface1">
    <g id="g14" style="fill:#000000;fill-opacity:1">
      <use height="100%" width="100%" id="use12" y="134.765" x="148.71201" xlink:href="#glyph0-1"/>
    </g>
  </g>
</svg>
```

Clearly, the units are interpreted as pt here and from the viewpoint of the API these are the user units of the snippet (The command line conversion pdf->svg always results in snippets the user units of which are in pt as it seems.) But they are not related to the document you are currently working with. 

What I do not understand is:
- ``tt_node.get_center_position()`` returns the position in pt
- ``elf.svg.get_center_position()`` returns the view center in, for example, mm

but when consecutively apply the transform we can mix the translate operations (one in pt, one mm).

Here is the situation in a debugging screenshot of the text ``Hello world!`, scale factor 1.0`:

![grafik](https://user-images.githubusercontent.com/11866252/71268906-44634880-234e-11ea-9105-e2680c1c3173.png)

You can reproduce it in a document of A4-size, fully centered in the screen (cx = 105mm, cy 148mm). The dimension of the snippet is w = 18.41mm, h=2.55mm or **in pt: w = 51.24pt, h = 7.25pt** the half of which matches the value in the screenshot of the debugging session.

Do you have any clue? (Obviously, this works, as we see, but why, that's my problem... :confused: )

Reply:
OK, I can give the answer for myself:

Node:
```bash
>> tt_node.transform
Transform(((1, 0, 0), (0, 1, 0)))
```` 

Move operation in pt:
```bash
>>Transform(translate=-node_center)*tt_node.transform
Transform(((1, 0, -25.7122), (0, 1, -3.62537)))
```

By applying the scalefactor the move operation is in fact in mm (in this example):
```bash
>>Transform(scale=scale_factor)*Transform(translate=-node_center)*tt_node.transform
Transform(((0.352778, 0, -9.07069), (0, 0.352778, -1.27895)))
```

And then the movment to the document center in mm
```bash
>>Transform(translate=view_center)*Transform(scale=scale_factor)*Transform(translate=-node_center)*tt_node.transform
Transform(((0.352778, 0, 96.7626), (0, 0.352778, 146.888)))
```

So this makes really sense!

In order to not rely on a hard-coded unit for the snippet maybe we should get the unit by
```python
# height="7.24pt" -> 'pt'
inkex.units.parse_unit(root.get('height'), 'pt', 1)[1]
```
directly after importing the snippet? Will fallback to 'pt' if no height attrib is found. What do you think?

Reply:
Ok. Now I see where 1pt came from! The right thing to do is to scale node at the import time to factor out snippet units. To find out ratio we can convert anything to user units of snippet and to user units of root document: 

```python
# pseudo-code
snippet_to_root_ratio = snippet.unittouu("1mm") / self.svg.unittouu("1mm")
node.transform.add_transform(1/snippet_to_root_ratio) 
```

It should be done in `TexTextElement._svg_to_textext_node()`. Root document will become constructor argument

Reply:
OK, I can give the answer for myself:

Node:
```bash
>> tt_node.transform
Transform(((1, 0, 0), (0, 1, 0)))
```` 

Move operation in pt:
```bash
>>Transform(translate=-node_center)*tt_node.transform
Transform(((1, 0, -25.7122), (0, 1, -3.62537)))
```

By applying the scalefactor the move operation is in fact in mm (in this example):
```bash
>>Transform(scale=scale_factor)*Transform(translate=-node_center)*tt_node.transform
Transform(((0.352778, 0, -9.07069), (0, 0.352778, -1.27895)))
```

And then the movment to the document center in mm
```bash
>>Transform(translate=view_center)*Transform(scale=scale_factor)*Transform(translate=-node_center)*tt_node.transform
Transform(((0.352778, 0, 96.7626), (0, 0.352778, 146.888)))
```

So this makes really sense!

In order to not rely on a hard-coded unit for the snippet maybe we should get the unit by
```python
# height="7.24pt" -> 'pt'
inkex.units.parse_unit(root.get('height'), 'pt', 1)[1]
```
directly after importing the snippet? Will fallback to 'pt' if no height attrib is found. What do you think?

Reply:
We were on the same track...

> Ok. Now I see where 1pt came from! The right thing to do is to scale node at the import time to factor out snippet units. To find out ratio we can convert anything to user units of snippet and to user units of root document

Yeah, that's the most elegant way. Now it is factored out in the consecutive transform operations. I will provide an implementation.

Reply:
This makes the scaling of the scale factor unnecessary, too. Things become consistent now...

Reply:
Yeah, units. I remember I had doubts about my modelling program and I spent days looking for origins of 2 pi factor... :crying_cat_face:

Reply:
FYI: I [proposed](https://gitlab.com/inkscape/extensions/merge_requests/157) removal of `.get_center_position()` methods from inkex API

The updated version should look like
```python
node_center = tt_node.bounding_box().center # works now
view_center = self.svg.namedview.center # not yet available 
```

Reply:
@jcwinkler All proposed changes are in master branch of `inkscape/extensions`. I don't know when they show up in appimages. May be you should delay changes until it happen. 

We may need finer control of inkscape version, because the change will be not compatible with alpha versions of inkscape.

Reply:
As far as I do understand the beta 2 Inkscape versions for Windows are continously updated from the master branch, the Linux AppImages not: https://inkscape.org/release/inkscape-1.0beta2/ .

Reply:
Center calculation needs to be changed to
```python
node_center = tt_node.bounding_box().center
view_center = self.svg.namedview.center
```